cli/src/merge_tools/external.rs

```diff
@@ -436,19 +441,38 @@ pub fn generate_diff(
         .map_err(ExternalToolError::SetUpDir)?;
     set_readonly_recursively(diff_wc.right_working_copy_path())
         .map_err(ExternalToolError::SetUpDir)?;
-    invoke_external_diff(ui, writer, tool, &diff_wc.to_command_variables())
+    invoke_external_diff(
+        ui,
+        writer,
+        tool,
+        diff_wc.temp_dir(),
+        &diff_wc.to_command_variables(),
+    )
 }
 
 /// Invokes the specified `tool` directing its output into `writer`.
 pub fn invoke_external_diff(
     ui: &Ui,
     writer: &mut dyn Write,
     tool: &ExternalMergeTool,
+    diff_dir: &Path,
     patterns: &HashMap<&str, &str>,
 ) -> Result<(), DiffGenerateError> {
     // TODO: Somehow propagate --color to the external command?
     let mut cmd = Command::new(&tool.program);
-    cmd.args(interpolate_variables(&tool.diff_args, patterns));
+    if tool.diff_do_chdir {
+        cmd.args(interpolate_variables(
+            &tool.diff_args,
+            &maplit::hashmap! {
+                "left" => file_util::relative_path(diff_dir, Path::new(patterns["left"])).to_str().unwrap().to_owned(),
+                "right" => file_util::relative_path(diff_dir, Path::new(patterns["right"])).to_str().unwrap().to_owned(),
```

Comment:
Can we make "left" and "right" in `patterns` already be relative instead?

Reply:
I think that's doable.

Reply:
Or... hm. Copying the files to the temp dir uses the absolute path, so I think we'd have to do the moral equivalent of this code in each caller anyway. I'll experiment a bit.

Reply:
The answer turns out to be "sort of". I've implemented using relative paths in `patterns` from the caller as a separate commit on top of the version that munges `patterns` from absolute to relative paths.

The change is labeled as "experimental" because I'm not super happy with it yet, but I wanted to get feedback on if this is the direction I should be going in.